### ISSUE TITLE
Awkward deletion order for workflow

### ISSUE THREAD

**Issue by GFriedrich** [spring-cloud/spring-cloud-app-broker#262]
Hi,
I'm wondering whether the current deletion order in the `AppDeploymentDeleteServiceInstanceWorkflow` is correct.
Currently if you create a new service instance, it will first create the backing services for you and afterwards create the backing apps.
When the service instance gets removed, the same order is used: Backing services first, backing apps second
To me this looks a bit strange because if you have a backing app which uses those backing services, the deletion will render the backing app broken for a moment until it gets deleted.
Therefore I would suggest to remove the backing apps first and afterwards delete the backing services.
Any objections? Otherwise I would open a PR for it.

**Comment by Albertoimpl** [comment]
Hi @GFriedrich, that is a very valid point.
It will be more correct the way you are proposing.
A PR is more than welcome.

Thanks!

**Comment by Albertoimpl** [comment]
Had to reopen this and revert the commit.
There was a reason why this was happening. The deletion of the service is async and it doesn't force delete.

If we invert the order we get the following error in our acceptance-tests:
```
  2019-08-21T16:22:57.64+0200 [APP/PROC/WEB/0] OUT 2019-08-21 14:22:57.643 DEBUG 15 --- [ry-client-nio-1] cloudfoundry-client.request              : DELETE https://api.cf.salmon.springapps.io:443/v2/spaces/55f43431-59d2-4978-a1e6-cd4c86063b6f
   2019-08-21T16:22:57.67+0200 [APP/PROC/WEB/0] OUT 2019-08-21 14:22:57.679 DEBUG 15 --- [ry-client-nio-2] cloudfoundry-client.response             : 400    https://api.cf.salmon.springapps.io:443/v2/spaces/55f43431-59d2-4978-a1e6-cd4c86063b6f (36 ms)
   2019-08-21T16:22:57.68+0200 [APP/PROC/WEB/0] OUT 2019-08-21 14:22:57.681 ERROR 15 --- [ry-client-nio-2] o.s.c.a.d.c.CloudFoundryAppDeployer      : Failed to undeploy app app-create-space-per1
   2019-08-21T16:22:57.68+0200 [APP/PROC/WEB/0] OUT org.cloudfoundry.client.v2.ClientV2Exception: CF-NonrecursiveSpaceDeletionFailed(290011): Resource inside space a3985afa-3beb-4cbf-8b54-230281ef293f must first be deleted, or specify recursive delete.
   2019-08-21T16:22:57.68+0200 [APP/PROC/WEB/0] OUT 	at org.cloudfoundry.reactor.util.ErrorPayloadMapper.lambda$null$0(ErrorPayloadMapper.java:46) ~[cloudfoundry-client-reactor-3.15.0.RELEASE.jar!/:na]
   2019-08-21T16:22:57.68+0200 [APP/PROC/WEB/0] OUT 	at org.cloudfoundry.reactor.util.ErrorPayloadMapper.lambda$null$9(ErrorPayloadMapper.java:98) ~[cloudfoundry-client-reactor-3.15.0.RELEASE.jar!/:na]
   2019-08-21T16:22:57.68+0200 [APP/PROC/WEB/0] OUT 	at reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:118) ~[reactor-core-3.2.8.RELEASE.jar!/:3.2.8.RELEASE]
```
Making it unable to be deleted.

Do you have an environment where you can run the acceptance tests?
You can find the instructions here:
https://github.com/spring-cloud/spring-cloud-app-broker/tree/master/spring-cloud-app-broker-acceptance-tests

Let me know if you have any questions.

**Comment by GFriedrich** [comment]
@Albertoimpl oh, that was unfortunate
I digged into the code and I think that deleting the space is already quite buggy (but maybe you can correct me here if I'm wrong):
The space gets deleted as soon as a backing app gets deleted at https://github.com/spring-cloud/spring-cloud-app-broker/blob/97fe3c31fab4f727627f8e26f479c1f2f38ae229/spring-cloud-app-broker-deployer-cloudfoundry/src/main/java/org/springframework/cloud/appbroker/deployer/cloudfoundry/CloudFoundryAppDeployer.java#L529
But what happens if you have two (or more) backing apps? According to https://github.com/spring-cloud/spring-cloud-app-broker/blob/97fe3c31fab4f727627f8e26f479c1f2f38ae229/spring-cloud-app-broker-core/src/main/java/org/springframework/cloud/appbroker/deployer/BackingAppDeploymentService.java#L60-L70 the deletion is executed in parallel.
So as soon as the deletion is executed in the `CloudFoundryAppDeployer` the first try will always fail. Once the second backing app is deleted (in case of two backing apps) the deletion will succeed and the space will be removed. Nevertheless an error will be logged for the first backing app. It's a quite odd behaviour.
I see two solutions here:
1. Simply adding the "deleteSpace" part to the deletion of service instances. And doing the same try and error until the very last service instance is deleted and the space deletion succeeds.
2. Doing a complete rewrite of creating and deleting spaces and put this in front of/behind creation/deletion of the whole app workflow.

What do you think?
(Again, please correct me, if I got something wrong...)

**Comment by GFriedrich** [comment]
@Albertoimpl I've created a new PR using solution 1 for now.
Maybe you can check whether the acceptance tests look better now.

**Comment by Albertoimpl** [comment]
Thanks for the PR @GFriedrich, this makes a difference and works well now. Nicely done!

I can still see a tiny exception in the logs but it is just because we fire the delete space right after firing the delete app operation. It won't block the deletion and the alternative would have been waiting for app deletion competition to delete the space

```
   2019-08-26T13:28:32.26+0200 [APP/PROC/WEB/0] OUT 2019-08-26 11:28:32.268 DEBUG 16 --- [ry-client-nio-4] cloudfoundry-client.response             : 200    https://api.cf.salmon.springapps.io:443/v2/apps/d2ec7ef8-dd65-48b2-9d3a-80f202b45a17/service_bindings?page=1 (100 ms)
   2019-08-26T13:28:32.27+0200 [APP/PROC/WEB/0] OUT 2019-08-26 11:28:32.274 DEBUG 16 --- [ry-client-nio-4] cloudfoundry-client.request              : DELETE https://api.cf.salmon.springapps.io:443/v2/apps/d2ec7ef8-dd65-48b2-9d3a-80f202b

 2019-08-26T13:28:32.97+0200 [APP/PROC/WEB/0] OUT 2019-08-26 11:28:32.972 DEBUG 16 --- [ry-client-nio-3] cloudfoundry-client.request              : DELETE https://api.cf.salmon.springapps.io:443/v2/spaces/3601bd93-762e-4c9f-b50f-519dfb513cfe
   2019-08-26T13:28:33.00+0200 [APP/PROC/WEB/0] OUT 2019-08-26 11:28:33.005 DEBUG 16 --- [ry-client-nio-4] cloudfoundry-client.response             : 400    https://api.cf.salmon.springapps.io:443/v2/spaces/3601bd93-762e-4c9f-b50f-519dfb513cfe (33 ms)
   2019-08-26T13:28:33.00+0200 [APP/PROC/WEB/0] OUT 2019-08-26 11:28:33.007 ERROR 16 --- [ry-client-nio-4] o.s.c.a.d.c.CloudFoundryAppDeployer      : Failed to undeploy app app-create-space-per1
   2019-08-26T13:28:33.00+0200 [APP/PROC/WEB/0] OUT org.cloudfoundry.client.v2.ClientV2Exception: CF-NonrecursiveSpaceDeletionFailed(290011): Resource inside space f980c0e1-11ed-4116-9c8e-9c3282e8bf26 must first be deleted, or specify recursive delete.
```

Will merge the PR. Thanks!
